### ISSUE TITLE
Assert.Contains(collection) fails in 2.5.1 when collection is an IQueryable<T> with Entity Framework - with work around/solution

### ISSUE THREAD

**Issue by PTAdvanced** [xunit/xunit#2791]
Today I upgraded from XUnit 2.5.0 to 2.5.1 and found a test failing, which had worked previously.

I downgraded back to XUnit 2.5.0 and the test worked as expected.

I diagnosed the issue and have determined a work around, which may help identify the solution (details below).

In my test I essentially have something along the lines:

```
IQueryable<MyType> query = GetQuery();

Assert.Contains(query, o => o.Key = 123);

var firstThing = query.First(); <- Test fails here.
```

Very simple. In my case the IQueryable<T> is actually an EF DBSet<T>, DbSet<T> implements IQueryable<T>.

The failure was:

```
System.InvalidOperationException: There is already an open DataReader associated with this Connection which must be closed first.
   at Microsoft.Data.SqlClient.SqlInternalConnectionTds.ValidateConnectionForExecute(SqlCommand command)
   at Microsoft.Data.SqlClient.SqlConnection.ValidateConnectionForExecute(String method, SqlCommand command)
   at Microsoft.Data.SqlClient.SqlCommand.ValidateCommand(Boolean isAsync, String method)
   at Microsoft.Data.SqlClient.SqlCommand.RunExecuteReader(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, TaskCompletionSource`1 completion, Int32 timeout, Task& task, Boolean& usedCache, Boolean asyncWrite, Boolean inRetry, String method)
   at Microsoft.Data.SqlClient.SqlCommand.RunExecuteReader(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, String method)
   at Microsoft.Data.SqlClient.SqlCommand.ExecuteReader(CommandBehavior behavior)
   at Microsoft.Data.SqlClient.SqlCommand.ExecuteDbDataReader(CommandBehavior behavior)
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteReader(RelationalCommandParameterObject parameterObject)
   at Microsoft.EntityFrameworkCore.Query.Internal.SingleQueryingEnumerable`1.Enumerator.InitializeReader(Enumerator enumerator)
   at Microsoft.EntityFrameworkCore.Query.Internal.SingleQueryingEnumerable`1.Enumerator.<>c.<MoveNext>b__21_0(DbContext _, Enumerator enumerator)
   at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded)
   at Microsoft.EntityFrameworkCore.Query.Internal.SingleQueryingEnumerable`1.Enumerator.MoveNext()
   at System.Linq.Enumerable.TryGetSingle[TSource](IEnumerable`1 source, Boolean& found)
   at lambda_method508(Closure, QueryContext)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](
```

**Solution**

To get the test working again I simply modified the line:

`Assert.Contains(query, o => o.Key = 123);`

to

`Assert.Contains(query.ToList(), o => o.Key = 123);`

The Assert.Contains method expects an IEnumerable<T>, and IQueryable<T> implements IEnumerable<T>. I couldn't find anything in the documentation which recommended using .ToList() on an IQueryable<T>

I assume some recent update in 2.5.1 in maybe the CollectionTracker or how the collection is being Wrapped is causing the underlying Entity Framework DataReader to be left Open, which then blocks any subsequent data reading through that instance of the DbContext.

**Comment by bradwilson** [comment]
The [`CollectionTracker` is disposed during `Assert.Contains`](https://github.com/xunit/assert.xunit/blob/f125ad27d024cbf78b6f77b1c38d182731ab4e47/CollectionAsserts.cs#L280). That [`Dispose` forwards onto our internal enumerator](https://github.com/xunit/assert.xunit/blob/f125ad27d024cbf78b6f77b1c38d182731ab4e47/Sdk/CollectionTracker.cs#L374) via [`DisposeInternal`](https://github.com/xunit/assert.xunit/blob/f125ad27d024cbf78b6f77b1c38d182731ab4e47/Sdk/CollectionTracker.cs#L738), which disposes the enumerator that we got from the collection.

This seems to be perfectly in line with the contract of enumeration: dispose the enumerator when you're done with it.

I don't _think_ it should be generally a problem with `IQueryable<T>` vs. `IEnumerable<T>`, since we treat it as the latter which means everything that's done on it (including enumeration, which is what we do) would cause the query to "collapse and execute". In point of fact, calling `.ToList()` on it does basically exactly that: enumerate it and copy it into a container, by way of [the `List<T>` constructor](https://github.com/dotnet/runtime/blob/01cab2988074d6bed31b0d5ea74c3902b4300205/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs#L82-L89).

So I guess I'm left with: what's going on with EF? Why isn't our pattern sufficient to "close to the query" when it's basically identical to what `ToList` does?

Out of curiosity, have you tried `2.5.2-pre.2` yet? Can you provide a repro project?

**Comment by PTAdvanced** [comment]
Hi Brad

I recreated the issue in my test with v 2.5.1 (i.e. without calling `.ToList()` on my `IQueryablee<T>`. I then updated to 2.5.2-pre.2, ran the test, and it passes.

Went back to the 2.5.1 stable release and the test fails.

I don't have a repro project right now, but I could try and put something together over the weekend if it will be helpful?

It looks like the issue is resolved in the upcoming version, but it's always nice to fully understand these issues for future reference isn't it.

**Comment by bradwilson** [comment]
I would be interesting to know what's going on, but if it's fixed, there's not much reason for you to dig into creating a repro project. There were two embedded collections issues in #2755, and [the second one was fixed after 2.5.1](https://github.com/xunit/assert.xunit/commit/03aec0e53d59c198abc9318b42fcbddc15bdcbbd). So it's going to come down to that, undoubtedly (some interaction where the inner collection wasn't getting handled properly, and since inner collections in ORMs tend to be cross-table relationships, that's probably what was keeping the connection alive).

I think there's no need for you to create a repro at this point. I'm going to mark this as a dup of (the second half) of #2755.
